pkg/blobcache/blobcache.go

```diff
@@ -0,0 +1,396 @@
+import (
+	"context"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/containers/image/docker/reference"
+	"github.com/containers/image/image"
+	"github.com/containers/image/manifest"
+	"github.com/containers/image/transports"
+	"github.com/containers/image/transports/alltransports"
+	"github.com/containers/image/types"
+	"github.com/containers/storage/pkg/ioutils"
+	digest "github.com/opencontainers/go-digest"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	// Transport caches written blobs and offers to return some from its cache.  Its
+	// name is "blob-cache", and references which are passed to it to be parsed
+	// should take the form "blob-cache:directory[,...]:reference", where _directory_
+	// is an optional directory list, and _reference_ is another reference.
+	// Importing this package registers Transport with the image library.
+	Transport                        = &blobCacheTransport{}
+	_         types.ImageReference   = &blobCacheReference{}
+	_         types.ImageSource      = &blobCacheSource{}
+	_         types.ImageDestination = &blobCacheDestination{}
+)
+
+type blobCacheTransport struct {
+}
+
+func init() {
+	transports.Register(Transport)
+}
+
+// BlobCache is an object which saves copies of blobs that are written to it while passing them
+// through to some real destination, and which can be queried directly in order to read them
+// back.
+type BlobCache interface {
+	types.ImageReference
+	// HasBlob checks if a blob that matches the passed-in digest (and size, if not -1), is
+	// present in the cache.
+	HasBlob(types.BlobInfo) (bool, int64, error)
+	// Directories returns the list of cache directories.
+	Directories() []string
+	// ClearCache() clears the contents of the cache directories.  Note that this also clears
+	// content which was not placed there by this cache implementation.
+	ClearCache() error
+}
+
+type blobCacheReference struct {
+	transport   *blobCacheTransport
+	reference   types.ImageReference
+	directories []string
+}
+
+type blobCacheSource struct {
+	reference *blobCacheReference
+	source    types.ImageSource
+}
+
+type blobCacheDestination struct {
+	reference   *blobCacheReference
+	destination types.ImageDestination
+}
+
+func makeFilename(blobSum digest.Digest, isConfig bool) string {
+	if isConfig {
+		return blobSum.String() + ".config"
+	}
+	return blobSum.String()
+}
+
+func (t *blobCacheTransport) Name() string {
+	return "blob-cache"
+}
+
+func (t *blobCacheTransport) ParseReference(reference string) (types.ImageReference, error) {
+	components := strings.SplitN(reference, ":", 2)
+	realRef := ""
+	directories := []string{}
+	switch len(components) {
+	case 2:
+		directories = strings.Split(components[0], ",")
+		if len(directories) == 1 && directories[0] == "" {
+			directories = []string{}
+		}
+		realRef = components[1]
+	default:
+		return nil, errors.Errorf("error parsing reference for %q image %q: bad syntax", t.Name(), reference)
+	}
```

Comment:
(Absolutely non-blocking: this would be shorter with an `if len(components) != 2 { /* return error */ }`, the two variable declarations outside the `switch` scope could be merged with the assignments inside the `switch`.)